Front-end developer test app submitted by Kiki Ren

kikiren opened the pull request:
- Use css grid approach for Masonry layout
- Responsive column count with media query
- Partial matching filter
- Welcome any feedback :)

Commit:
mvp

Commit:
first commit

Commit:
ready to push

kikiren committed:
Update README.md

Commit:
partial matching

Commit:
ready for pr

Commit:
ready for pr

flyingant commented:
Hi @kikiren , This is MaYi from WCL and reviewing your code.

it looks well on your implementation to the test. 
But I notice that you just loaded all images at once on the home page, would you like to challenge the idea about the lazy loading strategy or just implement it? (Like scroll to load more images?)

Thank you

Commit:
css approach doesn't work with lazy load, try masonry packaging

Commit:
now lazy loading

Commit:
infinite scroll

kikiren commented:
@flyingant, use intersection observer api to implement lazy loading and infinite scroll, due to the css grid approach, the first picture in each column besides the first column were lazy loaded as they were not intersected then the gallery page loaded, so I changed to a responsive masonry package.

johncalvinroberts commented on src/pages/GalleryPage.js:
```diff
@@ -0,0 +1,35 @@
+  const { filter } = useSelector((state) => state);
+  const { images } = useGetImages();
+  const dispatch = useDispatch();
+  useEffect(() => {
+    dispatch(setLoadmore(true));
+  }, []);
```
I'm actually getting a linting error from this line:
```
src/pages/GalleryPage.js
  Line 14:6: React Hook useEffect has a missing dependency: 'dispatch'. Either include it or remove the dependency array 
```

Do you know why eslint would warn us about this? How might you fix this?

johncalvinroberts commented on src/hooks/useGetImages.js:
```diff
@@ -0,0 +1,27 @@
+  const dispatch = useDispatch();
+  useEffect(() => {
+    if (loadMore) {
+      // default api
+      let API = "http://localhost:4000/pics";
```
If you were to use a "real" backend in production, how would you manage this variable?

johncalvinroberts commented on src/hooks/useGetImages.js:
```diff
@@ -0,0 +1,27 @@
+    if (loadMore) {
+      // default api
+      let API = "http://localhost:4000/pics";
+      axios
+        .get(API)
+        .then((res) => {
+          setImages([...images, ...res.data]);
+          dispatch(setLoadmore(false));
+        })
+        .catch((e) => {
+          console.log(e);
+        });
```
Can you update this to use `async/await` syntax?

johncalvinroberts reviewed (commented):
Hi Kiki. Great job on the frontend test. I left a few comments, feel free to address or change as you see fit.

Commit:
fixed ESLint error, use async/await, use env vars

kikiren commented:
@johncalvinroberts , 

1. In this case I just want useEffect to behave like componentDidMount , so I added  `// eslint-disable-next-line react-hooks/exhaustive-deps` to disable the eslint rule.
2. I will set env vars for apis as well as secret keys, updated in code.
3. Updated to async/await. :)
